### PR TITLE
Sip 15 liquidation tests

### PR DESCRIPTION
**Pull request by jacko125**
When opening a pull request to submit a new SIP, please use the suggested template: https://github.com/Synthetixio/SIPs/blob/master/sip-X.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing WIP PRs.
 - The build passes.
 - Your Github username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: SIPS/sip-15.md [hunk 1]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+
+- ✅ It succeeds and adds an liquidation Entry for Alice as flagged
+- ✅ It sets deadline as block.timestamp + liquidation delay of _2_ weeks.
+- ✅ It emits an event accountFlaggedForLiquidation(account, deadline)
+
+***
+
+When
+
+- Bob tries to flag Alice address for liquidation again
```
Comment (justinjmoses) [inline SIPS/sip-15.md:142]: Bob or anyone else

File: SIPS/sip-15.md [hunk 2]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+
+Then
+
+- ❌ It fails as Alice's address is already flagged
+
+***
+
+When
+
+- Alice does not fix her c ratio by burning sUSD back to the Liquidation Target ratio
```
Comment (justinjmoses) [inline SIPS/sip-15.md:152]: [np] These are really more `Given` steps - the final step is the `When`

File: SIPS/sip-15.md [hunk 3]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+
+When
+
+- Bob flags Alice address for liquidation
+
+Then
+
+- ✅ It succeeds and adds an liquidation Entry for Alice as flagged
+- ✅ It sets deadline as block.timestamp + liquidation delay of _2_ weeks.
+- ✅ It emits an event accountFlaggedForLiquidation(account, deadline)
```
Comment (justinjmoses) [inline SIPS/sip-15.md:136]: You can put these events in the interface above.

File: SIPS/sip-15.md [hunk 4]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+
+- ✅ Bob's sUSD balance is reduced by 100 sUSD, and Alice's SNX is transferred to Bob's address. The amount of SNX transferred is:
+- `100 sUSD / Price of SNX` = `100 sUSD / $1 = 100 SNX redeemed` + liquidation penalty `100 * 12% = 112 SNX` transferred to Bob.
+- Alice debt is reduced by 100 sUSD to `433.33 sUSD` and she has `688 SNX` remaining.
+
+***
+
+When
+
+- After Bob's liquidating 100 sUSD worth of SNX, Alice collateral ratio at 158.77% is still below the liquidation target ratio.
```
Comment (justinjmoses) [inline SIPS/sip-15.md:167]: [np] Given

File: SIPS/sip-15.md [hunk 5]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+- and Chad tries to liquidate Alice's SNX collateral with 50 sUSD
```
Comment (justinjmoses) [inline SIPS/sip-15.md:168]: Chad?! 🤣

File: SIPS/sip-15.md [hunk 6]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+
+Then
+
+- ✅ Bob's liquidation transaction only partially liquidates the `1000 sUSD` input.
+- The amount that is burned is calculated from the max of the required debt to bring Alice's collateral ratio to `300%`. Debt balance required to be at `300%` is `210.6666666667 sUSD`. The debt to burn is `383.33 - 210.6666666667 = 172.67 sUSD`
+- Bob's sUSD balance is reduced by `172.67 sUSD` and `172.67 + 10% = 189.937 SNX` is transferred to Bob.
+- Alice's debt balance is reduced to `383.33 - 172.67 sUSD = 210.66 sUSD` and she has `632 - 189.937 = 442.063 SNX` remaining, with a 209.85% collateral ratio.
+
+- ✅ Alice's liquidation entry is removed and returns false
+- ✅ An event is emitted that liquidation flag is removed for her address
```
Comment (justinjmoses) [inline SIPS/sip-15.md:192]: Please put this in the interface above as well

File: SIPS/sip-15.md [hunk 7]
```diff
@@ -112,8 +114,144 @@ The rationale for these liquidation mechanisms are:
...
+
+---
+
+When
+
+- Alice has been flagged for liquidation
+- and the price of SNX increases so her Collateral ratio is now above the liquidation target ratio
+- and she calls checkAndRemoveAccountInLiquidation
```
Comment (justinjmoses) [inline SIPS/sip-15.md:200]: This is the first time you've explained this function in the SIP - I suggest adding something on it elsewhere as well

File: SIPS/sip-15.md (hunk 3 above)
Comment (jacko125) [reply SIPS/sip-15.md:136]: I had the events in the Specifications section described in the functions before

File: SIPS/sip-15.md (hunk 7 above)
Comment (jacko125) [reply SIPS/sip-15.md:213]: Let me add description in the `Specs` section

File: SIPS/sip-15.md [hunk 8]
```diff
@@ -64,13 +80,17 @@ interface ILiquidations {
...
-    function setLiquidationRatio(uint ratio) external;
+    function setLiquidationRatio(uint _liquidationRatio) external;
 
-    function setLiquidationTargetRatio(uint target) external;
+    function setLiquidationPenalty(uint _penalty) external;
 }
 ```
+**Events**
+
+ - `accountFlaggedForLiquidation(address indexed account, uint deadline)`
```
Comment (justinjmoses) [inline SIPS/sip-15.md:92]: [np] CaptialCase for events jacko

File: SIPS/sip-15.md [hunk 9]
```diff
@@ -64,13 +80,17 @@ interface ILiquidations {
...
 
     // owner only
-    function setLiquidationDelay(uint time) external;
+    function setLiquidationDelay(uint _time) external;
 
-    function setLiquidationRatio(uint ratio) external;
+    function setLiquidationRatio(uint _liquidationRatio) external;
 
-    function setLiquidationTargetRatio(uint target) external;
+    function setLiquidationPenalty(uint _penalty) external;
```
Comment (justinjmoses) [inline SIPS/sip-15.md:87]: [np] underscoring the arguments is probably overkill

File: SIPS/sip-15.md [hunk 10]
```diff
@@ -66,11 +94,15 @@ interface ILiquidations {
     // owner only
     function setLiquidationDelay(uint time) external;
 
-    function setLiquidationRatio(uint ratio) external;
+    function setLiquidationRatio(uint liquidationRatio) external;
 
-    function setLiquidationTargetRatio(uint target) external;
+    function setLiquidationPenalty(uint penalty) external;
```
Comment (hav-noms) [inline SIPS/sip-15.md:99]: Update with `ISynthetix.liquidateDelinquentAccount(address account, uint susdAmount)`

Comment (jacko125) [reply SIPS/sip-15.md:99]: That is a good name

Comment (hav-noms) [reply SIPS/sip-15.md:99]: yeah makes it clear its for a bad actor not just anyone

**Comment by kaiynne** [comment]
Reviewed and made a few minor edits. One thing I would suggest is that given we are now raising the target ratio to 800% to be in line with the C ratio it would be helpful to modify the example as this will be most salient for people and allow them to more easily asses the impact and cost of being liquidated.

**Pull request merged**